tools/killswitch/src/killswitch.rs

```diff
@@ -0,0 +1,258 @@
+        let submitter_config = settings
+            .tx_submitters
+            .get(&channel.home)
+            .ok_or_else(|| Error::MissingTxSubmitterConf(channel.home.clone()))?;
+        chain_setup
+            .try_into_home(Some(submitter_config.clone()), None, None)
+            .await
+            .map_err(|report| Error::HomeInit(format!("{:#}", report)))
+    }
+
+    /// Make `ConnectionManagers` or return error
+    async fn make_connection_manager(
+        channel: &Channel,
+        settings: &Settings,
+    ) -> Result<ConnectionManagers> {
+        let setup_type = ChainSetupType::ConnectionManager {
+            remote_network: &channel.replica,
+        };
+        let chain_setup = Self::make_chain_setup(&channel.replica, setup_type, settings)?;
+        let submitter_config = settings
+            .tx_submitters
+            .get(&channel.replica)
+            .ok_or_else(|| Error::MissingTxSubmitterConf(channel.replica.clone()))?;
+        chain_setup
+            .try_into_connection_manager(Some(submitter_config.clone()), None)
+            .await
+            .map_err(|report| Error::ConnectionManagerInit(format!("{:#}", report)))
+    }
+
+    /// Make `Signers` or return error
+    async fn make_signer(channel: &Channel, settings: &Settings) -> Result<Signers> {
+        let config = settings
+            .attestation_signers
+            .get(&channel.home)
+            .ok_or_else(|| Error::MissingAttestationSignerConf(channel.home.clone()))?;
+        Signers::try_from_signer_conf(config)
+            .await
+            .map_err(|report| Error::AttestationSignerInit(format!("{:#}", report)))
+    }
+
+    /// Build a new `KillSwitch`, configuring best effort and storing, not returning most errors
+    pub(crate) async fn new(args: Args, settings: Settings) -> Result<Self> {
+        let channels = if args.all {
+            Self::make_channels(&settings)
+        } else {
+            let destination_network = args
+                .all_inbound
+                .expect("Should not happen. Clap requires this to be present");
+            let all = Self::make_channels(&settings);
+            Self::make_inbound_channels(&destination_network, all)
+        };
+        if channels.is_empty() {
+            // The one error we bail on, since there's nothing else left to do
+            return Err(Error::NoNetworks);
+        }
+
+        let futs = channels.into_iter().map(|channel| async {
+            let home_contract = Self::make_home(&channel, &settings).await;
+            let connection_manager = Self::make_connection_manager(&channel, &settings).await;
+            let attestation_signer = Self::make_signer(&channel, &settings).await;
+            ChannelKiller {
+                channel,
+                home_contract: Some(home_contract),
+                connection_manager: Some(connection_manager),
+                attestation_signer: Some(attestation_signer),
```

Comment:
Perhaps I'm missing something but it doesn't look like we need these objects to be `Option<Result<T>>` and can instead just use `Result<T>`? If they are always going to be `Some(Result<T>)` then I think we can ditch the `Option`

Reply:
This is a way to deal with struct mutability, needed because, among other things, `Error` can't be `Clone` (upstream errors are not). As we can `take()` an option, swapping in `None`, we can consume struct members without consuming the entire struct itself. You could refactor it to consume the structs in one go, but the readability would be arguably much worse.

Reply:
Using `mem::take` instead of `Option::take` may be more idiomatic here. Investigating

Reply:
I'm going to skip `mem::take` as the ergonomics are bad and it doesn't get along with `Result`